### PR TITLE
feat(linter): implement constructor-super rule

### LINKED ISSUE #2280: eslint/constructor-super
This is marked as implemented, but is currently just a stub implementation

### PR DESCRIPTION
**Pull request by taearls**
Closes #2280

I implemented the [constructor-super](https://eslint.org/docs/latest/rules/constructor-super) eslint rule. I think the implementation is pretty straightforward. I tried to keep it as simple as possible within the constraints of the rule. 

I'm open to any feedback you have on it!

I used Claude to help generate a high level overview description of the code changes and reviewed it myself for accuracy.

# Implementation Overview: `constructor-super` Rule

### Hybrid AST + CFG Approach

This implementation uses a sophisticated hybrid approach combining AST analysis with Control Flow Graph (CFG) traversal to achieve accurate detection of `super()` violations:

1. **AST-based Detection for LogicalExpressions**: CFG doesn't properly handle logical expressions like `a && super()`, so these are detected using AST traversal
2. **CFG-based Path Analysis**: Uses DFS traversal of the CFG to analyze all execution paths through the constructor
3. **Per-Path Validation**: Tracks `super()` calls across different execution paths to detect:
   - Missing `super()` calls (never called)
   - Conditional `super()` calls (not in all paths)
   - Duplicate `super()` calls (called multiple times)
   - Loop-based `super()` calls (invalid due to 0 or N executions)

### Key Features

- **Accurate duplicate detection**: Reports only the duplicate calls (2nd, 3rd, etc.), not the first occurrence, with spans sorted by source position
- **Complex control flow handling**: Properly handles try/catch, switches, loops, conditionals, and nested scopes
- **Invalid superclass detection**: Recognizes invalid superclass expressions like literals (`extends 5`), arithmetic assignments (`extends (B += C)`), etc.
- **Edge case support**: Handles ternary operators, logical expressions, early returns, and acceptable exits (throw/return with value)
- **ESLint compatibility**: Diagnostic messages match ESLint format exactly (including trailing periods)

### Superclass Classification

The rule classifies superclasses into four types:
- **None**: No extends clause
- **Null**: `extends null` (special case requiring explicit return)
- **Invalid**: Literals or invalid expressions (`extends 5`, `extends (B += C)`)
- **Valid**: Potentially valid class expressions

### Path Analysis Results

CFG path analysis produces four possible results per path:
- `NoSuper`: Path doesn't call `super()`
- `CalledOnce`: Path calls `super()` exactly once (valid)
- `CalledMultiple`: Path calls `super()` multiple times (duplicate)
- `ExitedWithoutSuper`: Path exits early via throw/return with value (acceptable)

## Test Coverage

- Comprehensive snapshot tests covering 40+ test cases
- Tests for all control flow patterns: if/else, switch, try/catch, loops, ternaries
- Edge cases: nested classes, arrow functions, logical expressions, invalid superclasses
- All diagnostic messages verified against ESLint's output format

## Changes

- Added `constructor_super.rs` with 873 lines of implementation
- Added snapshot tests with 310 lines covering all edge cases
- Registered rule in `rules.rs` and generated rule runner implementations
- Rule categorized as `nursery` (pending promotion after validation)

## Related

- ESLint rule: https://eslint.org/docs/latest/rules/constructor-super
- This is a **recommended** rule in ESLint's configuration

### CODE REVIEW

File: crates/oxc_linter/src/rules/eslint/constructor_super.rs [hunk 1]
```diff
@@ -0,0 +1,799 @@
...
+    ///             super();
+    ///         } else {
+    ///             super();
+    ///         }
+    ///     }
+    /// }
+    /// ```
+    ConstructorSuper,
+    eslint,
+    correctness
```
Comment (camc314) [inline crates/oxc_linter/src/rules/eslint/constructor_super.rs]:
```suggestion
    nursery
```

can we mark this as nursery for a release until we are certain of it's stability?

File: crates/oxc_linter/src/rules/eslint/constructor_super.rs [hunk 2]
```diff
@@ -0,0 +1,799 @@
...
+        }
+    }
+
+    /// Find all super() calls in the constructor body
+    fn find_super_calls(
+        _body: &oxc_ast::ast::FunctionBody,
+        ctx: &LintContext,
+        class_node_id: NodeId,
+    ) -> Vec<SuperCallInfo> {
+        ctx.nodes()
```
Comment (camc314) [inline crates/oxc_linter/src/rules/eslint/constructor_super.rs]: this'll be slow as it iterates over all nodes.

did you explore using the control graph to get the super calls, and decide whether they're reachable/not

Comment (taearls) [reply crates/oxc_linter/src/rules/eslint/constructor_super.rs]: I did not look into that, but I will check it out!

File: crates/oxc_linter/src/rules/eslint/constructor_super.rs (hunk 1 above)
Comment (taearls) [reply crates/oxc_linter/src/rules/eslint/constructor_super.rs]: done

File: crates/oxc_linter/src/rules/eslint/constructor_super.rs (hunk 2 above)
Comment (taearls) [reply crates/oxc_linter/src/rules/eslint/constructor_super.rs]: there I refactored the implementation to use CFG as much as possible. I did find that I needed to lean on some AST information in order to emit the correct diagnostic messages.

File: crates/oxc_linter/src/snapshots/eslint_constructor_super.snap [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+   ·                                              ───────
+   ╰────
+  help: Remove the duplicate 'super()' call
+
+  ⚠ eslint(constructor-super): Lacked a call of 'super()' in some code paths.
+   ╭─[constructor_super.tsx:1:21]
+ 1 │ class A extends B { constructor() { super() || super(); } }
+   ·                     ─────────────────────────────────────
+   ╰────
+  help: Ensure 'super()' is called in all code paths
```
Comment (camc314) [inline crates/oxc_linter/src/snapshots/eslint_constructor_super.snap]: this error doesn't look right? shouldn't it be that super ios called 2x?

Comment (taearls) [reply crates/oxc_linter/src/snapshots/eslint_constructor_super.snap]: ah good catch. I'll look into it!

Comment (camc314) [reply crates/oxc_linter/src/snapshots/eslint_constructor_super.snap]: thank you! this is amazing work!

File: crates/oxc_linter/src/snapshots/eslint_constructor_super.snap [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+   ╭─[constructor_super.tsx:3:20]
+ 2 │     
+ 3 │ ╭─▶                             constructor() {
+ 4 │ │                                   do {
+ 5 │ │                                       super();
+ 6 │ │                                   } while (foo);
+ 7 │ ╰─▶                             }
+ 8 │     
+   ╰────
+  help: Ensure 'super()' is called in all code paths
```
Comment (zzt1224) [inline crates/oxc_linter/src/snapshots/eslint_constructor_super.snap:302]: I think here should be duplicate super

File: crates/oxc_linter/src/snapshots/eslint_constructor_super.snap [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+  ⚠ eslint(constructor-super): Lacked a call of 'super()' in some code paths.
+   ╭─[constructor_super.tsx:1:21]
+ 1 │ class A extends B { constructor(a) { while (a) super(); } }
+   ·                     ─────────────────────────────────────
+   ╰────
+  help: Ensure 'super()' is called in all code paths
+
+  ⚠ eslint(constructor-super): Lacked a call of 'super()' in some code paths.
+   ╭─[constructor_super.tsx:1:21]
+ 1 │ class A extends B { constructor() { return; super(); } }
```
Comment (zzt1224) [inline crates/oxc_linter/src/snapshots/eslint_constructor_super.snap:251]: This should be "Add a 'super()' call to the constructor"? since the super() is never reached, there are only one code path

File: crates/oxc_linter/src/snapshots/eslint_constructor_super.snap (hunk 3 above)
Comment (taearls) [reply crates/oxc_linter/src/snapshots/eslint_constructor_super.snap]: fixed!

File: crates/oxc_linter/src/snapshots/eslint_constructor_super.snap (hunk 4 above)
Comment (taearls) [reply crates/oxc_linter/src/snapshots/eslint_constructor_super.snap:302]: this one seems to be correct according to the eslint tests https://github.com/eslint/eslint/blob/main/tests/lib/rules/constructor-super.js#L415

my understanding is that it's not a duplicate because the `do` block of a do/while loop should only run once, but it doesn't run in all code paths because if `foo` is false, then it will never run

File: crates/oxc_linter/src/snapshots/eslint_constructor_super.snap (hunk 5 above)
Comment (taearls) [reply crates/oxc_linter/src/snapshots/eslint_constructor_super.snap:251]: thanks for catching this! I just pushed up a fix.

File: crates/oxc_linter/src/rules/eslint/constructor_super.rs [hunk 6]
```diff
@@ -0,0 +1,921 @@
...
+                }
+            }
+        }
+    }
+}
+
+impl ConstructorSuper {
+    /// Check if constructor body contains a LogicalExpression with super()
+    /// CFG doesn't create proper instructions for logical expressions, so we check AST
+    /// Returns true only if super() appears in a truly conditional position (not both sides)
```
Comment (camc314) [inline crates/oxc_linter/src/rules/eslint/constructor_super.rs]: let me fix this

Comment (camc314) [reply crates/oxc_linter/src/rules/eslint/constructor_super.rs]: #14671

Comment (taearls) [reply crates/oxc_linter/src/rules/eslint/constructor_super.rs]: nice!!

File: crates/oxc_linter/src/snapshots/eslint_constructor_super.snap (hunk 4 above)
Comment (zzt1224) [reply crates/oxc_linter/src/snapshots/eslint_constructor_super.snap:302]: This is the ESLint test for it:

https://github.com/eslint/eslint/blob/ce40f74efd45f66d9fbfc6f78ce622ee72008485/tests/lib/rules/constructor-super.js#L513

The reason this is a duplicate is that when foo is true, the super call inside the do block will repeat itself.

Comment (taearls) [reply crates/oxc_linter/src/snapshots/eslint_constructor_super.snap:302]: Ah you're right my bad. I'll take a look. Thanks!

**Comment by camc314** [comment]
ugh i don't know why graphite closed this.

**Review by camc314 — APPROVED** [review]
Great work! thank you!

**Comment by overlookmotel** [comment]
From #479:

> 0/61 recommended rules are remaining as TODO

This was the last one. Full house! 🎉
